simtools/util/workflow_description.py

```diff
@@ -0,0 +1,571 @@
+            self.toplevel_meta['CTA']['PRODUCT']['DESCRIPTION'] = \
+                _user_meta['PRODUCT']['DESCRIPTION']
+            self.toplevel_meta['CTA']['PRODUCT']['CREATION_TIME'] = \
+                _user_meta['PRODUCT']['CREATION_TIME']
+            if 'CONTEXT' in _user_meta['PRODUCT']:
+                self.toplevel_meta['CTA']['PRODUCT']['CONTEXT'] = \
+                    _user_meta['PRODUCT']['CONTEXT']
+            if 'VALID' in _user_meta['PRODUCT']:
+                if 'START' in _user_meta['PRODUCT']['VALID']:
+                    self.toplevel_meta['CTA']['PRODUCT']['VALID']['START'] = \
+                        _user_meta['PRODUCT']['VALID']['START']
+                if 'END' in _user_meta['PRODUCT']['VALID']:
+                    self.toplevel_meta['CTA']['PRODUCT']['VALID']['END'] = \
+                        _user_meta['PRODUCT']['VALID']['END']
+            self.toplevel_meta['CTA']['PRODUCT']['ASSOCIATION'] = \
+                _user_meta['PRODUCT']['ASSOCIATION']
+            self.toplevel_meta['CTA']['PROCESS'] = _user_meta['PROCESS']
+        except KeyError:
+            self._logger.error("Error reading user input meta data")
+            raise
+
+        try:
+            self.workflow_config['PRODUCT']['FILENAME'] = os.path.splitext(
+                _user_meta['PRODUCT']['DATA'])[0]
+        except KeyError:
+            pass
+
+    def _fill_product_meta(self):
+        """
+        Fill metadata for data product
+
+        Raises
+        ------
+        KeyError
+            if relevant fields are not defined in toplevel metadata
+            dictionary
+
+        """
+
+        self.toplevel_meta['CTA']['PRODUCT']['ID'] = \
+            self.workflow_config['ACTIVITY']['ID']
+        self._logger.debug("Assigned ACTIVITE UUID {}".format(
+            self.toplevel_meta['CTA']['PRODUCT']['ID']))
+
+        try:
+            self.toplevel_meta['CTA']['PRODUCT']['FORMAT'] = \
+                self.product_data_file_format()
+        except KeyError:
+            self._logger.error("Error PRODUCT meta from user input meta data")
+            raise
+
+    def _fill_product_association_identifier(self):
+        """
+        Fill list of associations in top-level data model
+
+        Raises
+        ------
+        KeyError
+            if PRODUCT::ASSOCIATION is not found
+
+        """
+
+        try:
+            for association in self.toplevel_meta['CTA']['PRODUCT']['ASSOCIATION']:
+                association['ID'] = self._read_instrument_name(association)
+        except KeyError:
+            self._logger.error('Error reading PRODUCT:ASSOCIATION')
+            raise
+
+    def _read_instrument_name(self, association):
+        """
+        Returns a string defining the instrument following
+        the gammasim-tools naming convention derived from
+        PRODUCT:ASSOCIATION entry
+
+        """
+
+        try:
+            _instrument = \
```

Comment:
I think this kind of code should be a function in `names`. One where you provide the site, class, type and subtype and get the instrument. This won't be the only place where we need this.

Reply:
Moved this into a new function in `names`: `names:simtoolsInstrumentName` 54fd12b
(I think we need to review `names` at some point)